src/types/subst.rs

```diff
@@ -0,0 +1,82 @@
+use std::borrow::Cow;
+
+use crate::extension::ExtensionRegistry;
+
+use super::type_param::{TypeArg, TypeParam};
+use super::{Type, TypeBound, TypeRow};
+
+#[derive(Clone, Debug)]
+pub(crate) struct Substitution<'a> {
+    /// The number of variables bound locally-to the tree within which we are substituting
+    /// (i.e. beneath the point where we started applying the substitution),
+    /// so these should be untouched by the substitution
+    leave_lowest: usize,
+    /// What to do to variables that are affected
+    mapping: Mapping<'a>,
+}
+
+#[derive(Clone, Debug)]
+enum Mapping<'a> {
+    /// An explicit value to substitute for each bound variable
+    Values(Cow<'a, [TypeArg]>),
+    /// An amount to add to the index of any free variable - that is,
+    /// any free var, of index `i`, becomes the variable `(i +` this amount`)`
+    AddToIndex(usize),
+}
+
+impl<'a, T: Into<Cow<'a, [TypeArg]>>> From<T> for Substitution<'a> {
+    fn from(value: T) -> Self {
+        Self {
+            leave_lowest: 0,
+            mapping: Mapping::Values(value.into()),
+        }
+    }
+}
+
+impl<'a> Substitution<'a> {
+    pub(crate) fn apply_var(&self, idx: usize, decl: &TypeParam) -> TypeArg {
+        if idx < self.leave_lowest {
+            return TypeArg::new_var_use(idx, decl.clone());
+        }
+        match &self.mapping {
+            Mapping::Values(args) => args
+                .get(idx - self.leave_lowest)
+                .expect("Unexpected free type var")
+                .clone(),
+            Mapping::AddToIndex(diff) => TypeArg::new_var_use(idx + diff, decl.clone()),
+        }
+    }
+
+    pub(super) fn get_type(&self, idx: usize, bound: TypeBound) -> Type {
```

Comment:
Hmmm, maybe this should be `apply_type_var` to be more consistent with `apply_var`

Reply:
Or rename `apply_var` to `get_arg`?

Reply:
I've gone with `apply_to_var` and `apply_to_type_var`. If we later move all the `substitute` methods into the Substitution class, we can call those `apply_to_arg`, `apply_to_custom`, etc...